fix: js-waku / nwaku interop

richard-ramos opened the pull request:
Fixes #244 
Fixes #247 
Partially fixes #214

D4nte commented:
I tested 01f220ed1ff742c716430eb9dbb8928233aa94cf with js-waku:
I am getting

```
   Error: protocol selection failed
        at select (/home/fryorcraken/src/status-im/js-waku/node_modules/multistream-select/src/select.js:72:17)
        at processTicksAndRejections (node:internal/process/task_queues:96:5)
        at async Upgrader._multiplexOutbound (/home/fryorcraken/src/status-im/js-waku/node_modules/libp2p/src/upgrader.js:459:36)
        at async Upgrader.upgradeOutbound (/home/fryorcraken/src/status-im/js-waku/node_modules/libp2p/src/upgrader.js:204:44)
        at async ClassIsWrapper.dial (/home/fryorcraken/src/status-im/js-waku/node_modules/libp2p-websockets/src/index.js:54:18)
        at async TransportManager.dial (/home/fryorcraken/src/status-im/js-waku/node_modules/libp2p/src/transport-manager.js:111:14)
        at async /home/fryorcraken/src/status-im/js-waku/node_modules/libp2p/src/dialer/dial-request.js:81:18
        at async /home/fryorcraken/src/status-im/js-waku/node_modules/p-some/index.js:53:19
AggregateError: 
    Error: protocol selection failed
        at select (node_modules/multistream-select/src/select.js:72:17)
        at processTicksAndRejections (node:internal/process/task_queues:96:5)
        at async Upgrader._multiplexOutbound (node_modules/libp2p/src/upgrader.js:459:36)
        at async Upgrader.upgradeOutbound (node_modules/libp2p/src/upgrader.js:204:44)
        at async ClassIsWrapper.dial (node_modules/libp2p-websockets/src/index.js:54:18)
        at async TransportManager.dial (node_modules/libp2p/src/transport-manager.js:111:14)
        at async /home/fryorcraken/src/status-im/js-waku/node_modules/libp2p/src/dialer/dial-request.js:81:18
        at async /home/fryorcraken/src/status-im/js-waku/node_modules/p-some/index.js:53:19
    at maybeSettle (node_modules/p-some/index.js:31:11)
    at /home/fryorcraken/src/status-im/js-waku/node_modules/p-some/index.js:69:23
```

I am dialing with the following protocols:

```
  "/vac/waku/relay/2.0.0-beta2",
  "/vac/waku/relay/2.0.0",
```

What protocols are supported? I'd expect at least one to be selected.  

I am also getting some socket hang up issues:

```
  waku:nwaku nwaku args: --listen-address=127.0.0.1 --relay=true --rpc=true --rpc-admin=true --websocket-support=true --log-level=debug --tcp-port=8001 --rpc-port=8000 --websocket-port=8002 --store=false --filter=false --lightpush=false +0ms
  waku:nwaku nwaku 359751 started at 3:11:07 pm +32ms
  waku:nwaku Waiting to see 'Node setup complete' in nwaku logs +0ms
  waku:nwaku nwaku node has been started +35ms
  waku:nwaku RPC Query:  get_waku_v2_debug_v1_info [] +0ms
  waku:nwaku nwaku 359751 getting SIGINT at 3:11:07 pm +6ms
  waku:nwaku nwaku 359751 interrupted: true +0ms

FetchError: request to http://localhost:8000/ failed, reason: socket hang up
    at ClientRequest.<anonymous> (node_modules/node-fetch/lib/index.js:1461:11)
    at ClientRequest.emit (node:events:520:28)
    at ClientRequest.emit (node:domain:475:12)
    at Socket.socketOnEnd (node:_http_client:466:9)
    at Socket.emit (node:events:532:35)
    at Socket.emit (node:domain:475:12)
    at endReadableNT (node:internal/streams/readable:1346:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)
```

Port 8000 is rpc as you can see for the parameters.

Also getting other errors:

```
  waku:nwaku nwaku args: --listen-address=127.0.0.1 --relay=true --rpc=true --rpc-admin=true --websocket-support=true --log-level=debug --tcp-port=8004 --rpc-port=8003 --websocket-port=8005 --persist-messages=true +7ms
  waku:nwaku nwaku 360231 started at 3:11:24 pm +12ms
  waku:nwaku Waiting to see 'Node setup complete' in nwaku logs +0ms
  waku:nwaku nwaku 360215 process exited with 0 at 3:11:24 pm +1ms
  waku:nwaku nwaku node has been started +37ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '4d6573736167652030',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1654146684088000000
  }
] +0ms
  waku:nwaku RPC Response:  Response {
  size: 0,
  timeout: 0,
  [Symbol(Body internals)]: {
    body: PassThrough {
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 4,
      _maxListeners: undefined,
      _writableState: [WritableState],
      allowHalfOpen: true,
      [Symbol(kCapture)]: false,
      [Symbol(kCallback)]: null
    },
    disturbed: true,
    error: null
  },
  [Symbol(Response internals)]: {
    url: 'http://localhost:8003/',
    status: 400,
    statusText: 'Bad Request',
    headers: Headers { [Symbol(map)]: [Object: null prototype] },
    counter: 0
  }
} { result: null, error: 'illegal base64 data at input byte 16', id: 1 } +3ms
  waku:nwaku nwaku 360231 getting SIGINT at 3:11:24 pm +2ms
  waku:nwaku nwaku 360231 interrupted: true +0ms

AssertionError: expected null to be true
Expected :true
Actual   :[null]
<Click to see difference>

    at Context.<anonymous> (src/lib/waku_store/index.node.spec.ts:46:14)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

```

D4nte commented:
Feel free to test yourself:
1. checkout the js-waku `go-waku-drop-in branch`
2. Set some env variable:
```
export DEBUG=waku* # To get the JSON query/response for debugging
export WAKU_SERVICE_NODE_BIN=/home/fryorcraken/src/status-im/go-waku/build/waku # Point the go-waku bin
export WAKU_SERVICE_NODE_DIR=/home/fryorcraken/src/status-im/go-waku/build # May not be needed, folder where the go-waku bin will be executed
```
3. run the tests: `npm run test:unit`

richard-ramos commented:
```
Error: protocol selection failed
        at select (node_modules/multistream-select/src/select.js:72:17)
        at processTicksAndRejections (node:internal/process/task_queues:96:5)
        at async Upgrader._multiplexOutbound (node_modules/libp2p/src/upgrader.js:459:36)
        at async Upgrader.upgradeOutbound (node_modules/libp2p/src/upgrader.js:204:44)
        at async ClassIsWrapper.dial (node_modules/libp2p-websockets/src/index.js:54:18)
        at async TransportManager.dial (node_modules/libp2p/src/transport-manager.js:111:14)
        at async /home/fryorcraken/src/status-im/js-waku/node_modules/libp2p/src/dialer/dial-request.js:81:18
        at async /home/fryorcraken/src/status-im/js-waku/node_modules/p-some/index.js:53:19
 ```
This error happened because go-libp2p v0.19.0 removed mplex from the list of default muxers (which is what nim-libp2p and js-libp2p support at the moment). I added it back in `master` but forgot to rebase this branch. See https://github.com/libp2p/specs/pull/402 and https://github.com/libp2p/go-libp2p/pull/1344

---

```
FetchError: request to http://localhost:8000/ failed, reason: socket hang up
```
Fixed! 

---

We're getting better results now :)
```
  107 passing (1m)
  1 pending
  24 failing
```

I'll have a look at the `illegal base64 data at input byte 16` error, but there are some others that I need some help understanding them: https://gist.github.com/richard-ramos/147ac70ff0ec80dcfaf101eb2a198db6

alaibe commented on waku.go:
```diff
@@ -111,6 +111,16 @@ func main() {
+			},
+			&cli.StringFlag{
+				Name:        "nat",
+				Usage:       "TODO",
```
Maybe add a better description even if todo?

alaibe reviewed (approved):
code LGTM
